src/Microsoft.DocAsCode.Build.Engine/DocumentBuildContext.cs

```diff
@@ -193,6 +199,81 @@ private List<string> ResolveByExternalReferencePackages(List<string> uidList, Co
             return list;
         }
 
+        private async Task<List<string>> ResolveByXRefServiceAsync(List<string> uidList, ConcurrentDictionary<string, XRefSpec> externalXRefSpec)
+        {
+            if (_xrefServiceUrls == null || _xrefServiceUrls.Length == 0)
+            {
+                return uidList;
+            }
+
+            var unresolvedUidList = new List<string>();
+            using (var client = new HttpClient())
+            {
+                client.DefaultRequestHeaders.Accept.Clear();
+                client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+                int pieceSize = 1000;
+                for(int i = 0; i < uidList.Count; i += pieceSize)
+                {
+                    List<string> smallPiece = uidList.GetRange(i, Math.Min(pieceSize, uidList.Count - i));
+                    foreach(string requestUrl in _xrefServiceUrls)
+                    {
+                        var queryUidTasks = new List<Task<List<XRefSpec>>>();
+                        foreach(string uid in smallPiece)
+                        {
+                            queryUidTasks.Add(QueryByHttpRequestAsync(client, requestUrl, uid));
+                        }
+                        var results = await Task.WhenAll(queryUidTasks);
+                        List<string> stillNeedQueryList = new List<string>();
+                        for(int j = 0; j < results.Length; j++)
+                        {
+                            if(results[j].Count > 0 && results[j][0].Uid == smallPiece[j])
+                            {
+                                externalXRefSpec.AddOrUpdate(smallPiece[j], results[j][0], (_, old) => old + results[j][0]);
+                            }
+                            else
+                            {
+                                stillNeedQueryList.Add(smallPiece[j]);
+                            }
+                        }
+                        smallPiece = stillNeedQueryList;
+                    }
+                    unresolvedUidList.AddRange(smallPiece);
+                }
+            }
+            Logger.LogInfo($"{uidList.Count - unresolvedUidList.Count} uids found in xrefservice.");
+            return unresolvedUidList;
+        }
+
+        internal async Task<List<XRefSpec>> QueryByHttpRequestAsync(HttpClient client, string requestUrl, string uid)
+        {
+            string url = requestUrl.Replace("{uid}", Uri.EscapeDataString(uid));
+            List<XRefSpec> emptyList = new List<XRefSpec>();
```

Comment:
instead ,change to return null

---
In reply to: [132073842](https://github.com/dotnet/docfx/pull/1837#discussion_r132073842) [](ancestors = 132073842)